### PR TITLE
Kick refactor

### PR DESCRIPTION
**Pull request by grandizzy**
based on Ian's https://github.com/ajna-finance/contracts/pull/258 PR

- added Auctions / AuctionsQueue library: kick starts a liquidation (Liquidation struct), remove delete the liquidation from queue
- gas load tests
- updated kick tests

gas costs for kicks
```
│ Function Name     ┆ min         ┆ avg        ┆ median ┆ max        ┆ # calls   │
│ kick                       ┆ 211117   ┆ 242795 ┆ 238420 ┆ 1060341 ┆ 16000   │
```

PR contains a (temporary) workaround for contract deployment size - move all PRB related function in external lib (approach adds about 1200 more gas due to delegate call )
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  22,233B  (90.46%)
  ERC20Pool                -  20,396B  (82.99%)
```

### CODE REVIEW

File: src/libraries/BucketMath.sol [hunk 1]
```diff
@@ -108,4 +112,42 @@ library BucketMath {
...
     }
 
+    function auctionPrice(
+        uint256 referencePrice,
+        uint256 kickTime_
+    ) public view returns (uint256 price_) {
+        uint256 elapsedHours = Maths.wdiv((block.timestamp - kickTime_) * 1e18, 1 hours * 1e18);
+        elapsedHours -= Maths.min(elapsedHours, 1e18);  // price locked during cure period
+
+        int256 timeAdjustment = PRBMathSD59x18.mul(-1 * 1e18, int256(elapsedHours));
```
Comment (mattcushman) [inline src/libraries/BucketMath.sol:122]: slight gas savings, is it possible to do the subtraction in the opposite way (line 119) and then not multiply by -1?

File: src/libraries/Auctions.sol [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        address head;
+        address tail;
+        mapping(address => Liquidation) liquidations;
+    }
+
+    struct Liquidation {
+        address kicker;     // address that initiated liquidation
+        uint256 bondSize;   // bond size posted by kicker to start liquidation
+        uint256 bondFactor; // bond factor used to start liquidation
+        uint128 time;       // timestamp when liquidation was started
```
Comment (ith-harvey) [inline src/libraries/Auctions.sol]: I like the naming convention of `kickTime` better personally.

File: src/libraries/Auctions.sol [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        uint128 hpbIndex;   // HPB index at liquidation start
```
Comment (ith-harvey) [inline src/libraries/Auctions.sol]: Like the naming convention of `kickPriceIndex` better here.

File: src/libraries/Auctions.sol [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        address prev;       // previous liquidated borrower in auctions queue
+        address next;       // next liquidated borrower in auctions queue
```
Comment (ith-harvey) [inline src/libraries/Auctions.sol]: Nit: all first letter of each comment should be capitalized in the section above ^^

File: src/libraries/Auctions.sol [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+        liquidation.next = address(0);
+        if (self_.head != address(0)) {
+            // other auctions in queue, liquidation doesn't exist or overwriting.
+            self_.liquidations[self_.tail].next = borrower_;
+            liquidation.prev = self_.tail;
+        } else {
+            // first auction in queue
+            self_.head = borrower_;
+            liquidation.prev  = address(0);
```
Comment (ith-harvey) [inline src/libraries/Auctions.sol:79]: Appears you deceided to add the `delete` call in remove. That should work fine to ensure auctions who are removed and re added to the auction queue don't have stale pointers :thumbs_up:

File: src/libraries/Auctions.sol [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    /**************************/
+    /*** View Functions ***/
+    /**************************/
+
+    function getHead(Data storage self_) internal view returns (address) {
+        return self_.head;
+    }
+
+    function isActive(Data storage self_, address borrower_) internal view returns (bool) {
+        return self_.liquidations[borrower_].kicker != address(0);
```
Comment (ith-harvey) [inline src/libraries/Auctions.sol:129]: really like this!

File: src/libraries/Book.sol [hunk 7]
```diff
@@ -181,16 +181,22 @@ library Book {
         }
     }
 
+    function momp(
```
Comment (ith-harvey) [inline src/libraries/Book.sol:184]: This needs natspec but I can add in seperate PR

File: src/libraries/BucketMath.sol [hunk 8]
```diff
@@ -108,4 +112,42 @@ library BucketMath {
...
+        uint256 lastInflatorSnapshotUpdate_,
+        uint256 interestRate_
+    ) public view returns (uint256) {
+        return Maths.wmul(
+            inflatorSnapshot_,
+            PRBMathUD60x18.exp((interestRate_ * (block.timestamp - lastInflatorSnapshotUpdate_)) / 365 days)
+        );
+    }
+
+    function lenderInterestMargin(
```
Comment (ith-harvey) [inline src/libraries/BucketMath.sol:144]: I'm not the biggest fan of including these methods here: `lenderInterestMargin`, `pendingInflator`, `auctionPrice`, `pendingInterestFactor` Don't think they belong un bucket math as they are more of a pool based concepts

File: src/base/interfaces/pool/IPoolState.sol [hunk 9]
```diff
@@ -7,6 +7,27 @@ pragma solidity 0.8.14;
...
+     *  @return hpbIndex   Highest Price Bucket index at time of liquidation.
+     *  @return prev       The address of previous borrower in auctions queue.
+     *  @return next       The address of next borrower in auctions queue.
+     */
+    function auctionInfo(address borrower) external view returns (
+        address kicker,
+        uint256 bondSize,
+        uint256 bondFactor,
+        uint128 kickTime,
+        uint128 hpbIndex,
```
Comment (ith-harvey) [inline src/base/interfaces/pool/IPoolState.sol]: bigger fan of `kickPriceIndex`

File: src/base/Pool.sol [hunk 10]
```diff
@@ -340,9 +343,59 @@ abstract contract Pool is Clone, Multicall, IPool {
...
+       ) revert BorrowerOk();
+
+        uint256 thresholdPrice = borrowerAccruedDebt * Maths.WAD / borrowerPledgedCollateral;
+        if (lup > thresholdPrice) revert LUPGreaterThanTP();
+
+        uint256 bondSize = auctions.kick(
+            borrower_,
+            borrowerAccruedDebt,
+            thresholdPrice,
+            deposits.momp(poolState.accruedDebt, loans.nodes.length - 1),
```
Comment (ith-harvey) [inline src/base/Pool.sol:378]: This actually needs to be the current mompFactor of the loan  multiplied by the inflator (reverse of how we store the `mompFactor`).  So this line should be `Maths.wmul(borrower.mompFactor, borrower.inflator)`

Obviously need to grab the right borrower data to perform this calc ^^

**Review by ith-harvey — CHANGES_REQUESTED** [review]
Here you go @gniculae , as I said earlier looks really good. Happy to approve once we make these changes.

File: src/libraries/BucketMath.sol (hunk 8 above)
Comment (grandizzy) [reply src/libraries/BucketMath.sol:144]: yeah, that's just a temporary workaround to get us going

File: src/libraries/Auctions.sol (hunk 4 above)
Comment (EdNoepel) [reply src/libraries/Auctions.sol]: Disagree; these are not sentences

File: src/base/Pool.sol (hunk 10 above)
Comment (grandizzy) [reply src/base/Pool.sol:378]: there are 2 functions in Book library - `momp` and `mompFactor` https://github.com/ajna-finance/contracts/blob/kick-refactor/src/libraries/Book.sol#L184 I don't think that's an issue?

File: src/libraries/Auctions.sol (hunk 2 above)
Comment (grandizzy) [reply src/libraries/Auctions.sol]: renamed with https://github.com/ajna-finance/contracts/pull/283/commits/2a4ecafbb4fb8fd83d91dc4f8cb37713ba91236d

File: src/libraries/Auctions.sol (hunk 3 above)
Comment (grandizzy) [reply src/libraries/Auctions.sol]: renamed with https://github.com/ajna-finance/contracts/pull/283/commits/2a4ecafbb4fb8fd83d91dc4f8cb37713ba91236d

File: src/base/interfaces/pool/IPoolState.sol (hunk 9 above)
Comment (grandizzy) [reply src/base/interfaces/pool/IPoolState.sol]: renamed with https://github.com/ajna-finance/contracts/pull/283/commits/2a4ecafbb4fb8fd83d91dc4f8cb37713ba91236d

File: src/base/Pool.sol (hunk 10 above)
Comment (ith-harvey) [reply src/base/Pool.sol:378]: This is correct as is, my mistake. Resolving.

File: src/libraries/BucketMath.sol (hunk 8 above)
Comment (ith-harvey) [reply src/libraries/BucketMath.sol:144]: Ok... should they exist in pool utils or pool then? Or you want to keep them here?

File: src/libraries/Auctions.sol (hunk 4 above)
Comment (ith-harvey) [reply src/libraries/Auctions.sol]: As it's not a convention i.e. we don't follow that rule elsewhere I agree fine as is.

File: src/libraries/BucketMath.sol (hunk 8 above)
Comment (grandizzy) [reply src/libraries/BucketMath.sol:144]: Should be in PoolUtils library and BucketMath retired entirely. But have to shrink contract size somehow

File: src/libraries/Auctions.sol (hunk 4 above)
Comment (grandizzy) [reply src/libraries/Auctions.sol]: Yeah, they're not natspecs but rather internal  comments. The natspec in pool interface start all lines with capital letter

Comment (EdNoepel) [reply src/libraries/Auctions.sol]: Convention in the natspec is to define terms in a sentence-like structure which ends in a period and is suitable for doc autogeneration.  Convention outside of the natspec is to use lowercase definitions which are not terminated by a period.

**Review by ith-harvey — APPROVED** [review]
LGTM!

**Pull request merged**